### ISSUE TITLE
some probelm in your manuscript

### ISSUE THREAD

**Issue by ShaowuChen** [MingSun-Tse/Why-the-State-of-Pruning-so-Confusing#2]
Very good work! Very helpful to the community!

Check the 2nd paragraph in page 5:
1. Scratch-E and Scratch-B are from RETHINKING THE VALUE OF NETWORK PRUNING, not from Pruning filters for efficient convnets.
2. From RETHINKING THE VALUE OF NETWORK PRUNING, it seems that Scratch-B takes K1*F1/F2 epochs , instead of K1*F1/F2+K2. (I am not 100% sure about this.)

**Comment by MingSun-Tse** [comment]
Hi @ShaowuChen , thanks for your interest in our work!

1. You are right. The reference is wrong. We shall correct this in the next version. Thank you for pointing this out!
2. `K1F1/F2 + K2` is a more precise version of "Scratch-B". The rethinking paper omits the `+ K2` part (hence `K1F1/F2`) because their finetuning **does not take long**; see the following text I copy from the rethinking paper:

> Note that our evaluations use the same computation as large model training _without considering the computation in fine-tuning, since in our evaluated methods fine-tuning does not take too long_; if anything this still favors the pruning and fine-tuning pipeline. 

Since in our experiments, the finetuning does take long, much longer than what Liu et al. did, using `K1F1/F2 + K2` is more correct.

Let me know if you have more questions! Thank you for the discussion!

**Comment by ShaowuChen** [comment]
Thank you for your clear explanations.

**Comment by ShaowuChen** [comment]
Hello @MingSun-Tse ！

1. Two minor typos in your manuscript:
In the last paragraph in the left column of page 10:
 (1) Tab. 2->Tab. 5 (See also in the next paragraph)
 (2) The test accuracy "76.74" is not existed in Tab. 5.  (Maybe "77.81"？ Or "77.81"->"76.74"?)

2. I have 2 questions: 
 (1) As stated in the end of page 7 and the start of page 8:
> we are not allowed to change the LR schedule. The only thing we can change is the pruning epoch. It is confused to me, since the LR schedules  in the blue-marked group are different from each other.

It is confused to me, since the LR schedules (i.e., inital LR) in the blue-marked group are different from each other. And it seems that groups mared by blue and yellow are all belong to the same group, SX-A. Do I misunderstand something? 

(2) The reimplemented L1 [42] borrows the layerwise pruning ratio from GReg [79]. In this scenario, the outperformance of the reimplemented L1 may mostly come from architechure searched by GReg, instead of the appropriate LR schedule. I am curious about how the orignal L1 [42] (under different and uniform layerwise ratio, respectively) will perform if with the optimal LR schudule (i.e., large LR, more epochs). Furthermore, random pruning under the same settings (e.g., ratio, LR) is also important to show the value of pruning. Can you add them to your paper? I think they would greatly unveil pruning mysteries.

**Comment by MingSun-Tse** [comment]
Hi @ShaowuChen , thanks again for your interest and discussion!

1. You are right about the "Tab. 2" typo. Thanks for letting us know! 

I double-checked the results of P60F60, finetuning LR0.01, and confirm that 76.74 should be updated to 77.81, see our log below.
![image](https://user-images.githubusercontent.com/16812048/219051045-c8d86df1-b856-43f1-af54-0c6c2497cfa7.png)
(76.74 should be only an intermediate result when I wrote the paper, yet forgot to update it after..)

Thanks for pointing these out! we shall update the paper and upload it to this repo shortly, before we finally update it on arxiv.

2. (1) 
> It is confused to me, since the LR schedules (i.e., inital LR) in the blue-marked group are different from each other. And it seems that groups mared by blue and yellow are all belong to the same group, SX-A. Do I misunderstand something?

For the blue rows, we maintain the _overall_ (pretraining+pruning+finetuning, since L1-norm pruning spends no epochs in pruning, so this reduces to pretraining+finetuning) LR schedule the same. The initial LR in the parentheses is _for finetuning_. If you check the overall LR schedule, you would notice the blue rows are different from the orange rows. Let me give you a concrete example:

For the yellow rows, take `L1-norm (P30F90, 1e-1)` as an example, its finetuning LR schedule is: `0:1e-1, 30:1e-2, 60:1e-3, 75:1e-4, 83:1e-5, total: 90 epochs` (see the Appendix A for this). Plus the pertaining stage (LR 0.1), this makes the overall LR schedule be: `0:1e-1, 30:1e-1, 60:1e-2, 90:1e-3, 105:1e-4, 113:1e-5, total: 120 epochs` -- this is different from the LR schedule used by scratch training and the blue rows, which is, `Initial LR 0.1, decayed at epoch 30/60/90/105 by multiplier 0.1, total: 120 epochs` (see the caption of Tab. 4).

So as seen, if we want to compare L1-norm pruning with Scratch training, the blue rows are in a _more strictly controlled_ condition than the orange ones, and possibly reveal a more faithful result.  The theme of this paper is to show, the current comparison setups (<=S3.1) used by the community have many loopholes. Even if we meet them (like, the blue and orange rows both meet <=S3.1), we can still obtain very different results, leading to very different conclusions. Therein, we might need to think over which setup is closer to the truth?

(2) I see your point and we will update the manuscript per your suggestions after we have new results (like the random pruning). Thanks! To quickly answer your questions here:

- (First, GReg does not really "search" the layerwise ratios; the ratios are manually chosen simply by a few trials and errors with a given speedup.)
- Then, for the results in Tab. 1, the strong performance of L1 is attributed to the smart layerwise pruning ratio, or, the better LR schedule? I think the layerwise ratio definitely contributes as we know uniform pruning typically does not work that well. Meanwhile, the better finetuning also plays an important part -- Check out Tab. 5 of this paper (https://arxiv.org/pdf/2105.05916.pdf, a quite early version of this paper), where uniform layerwise pruning ratio is used and you can see the larger finetuning LR still (mostly) boosts the performance significantly. Note, we do not take any credit for the performance-boosting effect of a larger finetuning LR, because it is _not_ our finding. At least, two papers, [37, 65], can support this phenomenon is true.

Let me know if you have any more questions or suggestions. Regards,

**Comment by ShaowuChen** [comment]
Thank you again for your detailed answer!
